mldsa/sign.c

```diff
@@ -14,6 +14,81 @@
 #include "sign.h"
 #include "symmetric.h"
 
+
+static int mld_check_pct(uint8_t const pk[CRYPTO_PUBLICKEYBYTES],
+                         uint8_t const sk[CRYPTO_SECRETKEYBYTES])
+__contract__(
+  requires(memory_no_alias(pk, CRYPTO_PUBLICKEYBYTES))
+  requires(memory_no_alias(sk, CRYPTO_SECRETKEYBYTES))
+);
+
+#if defined(MLD_CONFIG_KEYGEN_PCT)
+/*************************************************
+ * [FIPS 140-3 IG]
+ * (https://csrc.nist.gov/csrc/media/Projects/cryptographic-module-validation-program/documents/fips%20140-3/FIPS%20140-3%20IG.pdf)
+ *
+ * TE10.35.02: Pair-wise Consistency Test (PCT) for DSA keypairs
+ *
+ * Purpose: Validates that a generated public/private key pair can correctly
+ * sign and verify data. Test performs signature generation using the private
+ * key (sk), followed by signature verification using the public key (pk).
+ * Returns 0 if the signature was successfully verified, non-zero if it cannot.
+ *
+ * Note: FIPS 204 requires that public/private key pairs are to be used only for
+ * the calculation and/of verification of digital signatures.
+ **************************************************/
+static int mld_check_pct(uint8_t const pk[CRYPTO_PUBLICKEYBYTES],
+                         uint8_t const sk[CRYPTO_SECRETKEYBYTES])
+{
+  uint8_t message[1] = {0};
+  uint8_t signature[CRYPTO_BYTES];
+  uint8_t pk_test[CRYPTO_PUBLICKEYBYTES];
+  size_t siglen;
+  int ret;
+
+  /* Copy public key for testing */
+  memcpy(pk_test, pk, CRYPTO_PUBLICKEYBYTES);
+
+  /* Sign a test message using the original secret key */
+  ret = crypto_sign_signature(signature, &siglen, message, sizeof(message),
+                              NULL, 0, sk);
+  if (ret != 0)
```

Comment:
I am aesthetically opposed to the use of "goto" here... especially since there's only one of them. Wouldn't an simple "if" statement here suffice?